src/search/parser_tests.h

```diff
@@ -0,0 +1,210 @@
+        std::cout << indent << "searches = " << std::endl;
+        for (const auto& search_AST : searches) {
+            parser::Any search_any = search_AST->construct();
+            std::shared_ptr<SearchEngine> search = parser::any_cast<std::shared_ptr<SearchEngine>>(search_any);
+            search->dump(indent + "    ");
+        }
+    }
+};
+
+class IteratedSearchPlugin : public parser::Plugin<SearchEngine, IteratedSearch> {
+public:
+    IteratedSearchPlugin()
+        : parser::Plugin<SearchEngine, IteratedSearch>("iterated", "Iterated Search") {;
+        add_option<std::vector<std::shared_ptr<SearchEngine>>>(
+            "searches",
+            "use preferred operators of these evaluators", "[]", parser::Bounds::unlimited(), true);
+    }
+};
+
+class BlindHeuristicPlugin : public parser::Plugin<Evaluator, BlindHeuristic> {
+public:
+    BlindHeuristicPlugin()
+        : parser::Plugin<Evaluator, BlindHeuristic>("blind", "Blind Heuristic") {
+        document_synopsis("Returns cost of cheapest action for "
+                          "non-goal states, "
+                          "0 for goal states");
+        document_language_support("action costs", "supported");
+        document_language_support("conditional effects", "supported");
+        document_language_support("axioms", "supported");
+        document_property("admissible", "yes");
+        document_property("consistent", "yes");
+        document_property("safe", "yes");
+        document_property("preferred operators", "no");
+
+        add_option<int>("repeat", "how often to check for the goal", "2K");
+        add_option<double>("time_limit", "when to stop", "2e4");
+        add_list_option<std::vector<int>>("patterns", "...", "[[1,2,3], []]");
+
+        std::vector<std::string> lp_solvers;
+        std::vector<std::string> lp_solvers_doc;
+        lp_solvers.push_back("CLP");
+        lp_solvers_doc.push_back("default LP solver shipped with the COIN library");
+        lp_solvers.push_back("CPLEX");
+        lp_solvers_doc.push_back("commercial solver by IBM");
+        lp_solvers.push_back("GUROBI");
+        lp_solvers_doc.push_back("commercial solver");
+        lp_solvers.push_back("SOPLEX");
+        lp_solvers_doc.push_back("open source solver by ZIB");
+        add_enum_option<LPSolverType>(
+            "lpsolver",
+            lp_solvers,
+            "external solver that should be used to solve linear programs",
+            "CPLEX",
+            lp_solvers_doc);
+    }
+
+    virtual parser::Options preprocess_options(const parser::Options &options) const override {
+        options.verify_list_non_empty<std::vector<int>>("patterns");
+        parser::Options copy(options);
+        std::vector<std::vector<int>> patterns = options.get_list<std::vector<int>>("patterns");
+        for (std::vector<int> &pattern : patterns) {
+            pattern.push_back(42);
+        }
+        copy.set("patterns", patterns);
+        return copy;
```

Comment:
The method would be used for two things in the current plugins:
1) Testing the validity of options after parsing. For example, some lists may not be empty, some options can only be set if other optional options are set as well.  This is code that is currently called after `Options opts = parser.parse();` and before returning the constructed object. For examples, see `src/search/evaluators/sum_evaluator.cc` or `src/search/merge_and_shrink/label_reduction.cc`.
2) Creating new `Options` objects or modifying the existing one. Some plugins currently create an object with the given options and then create a new `Options` object that contains the newly created object and use the new options to create the final plugin. For example, parsing `ipdb()` uses the options to create a `PatternCollectionGeneratorHillclimbing` object, then creates a new options object and uses it to create a `CanonicalPDBsHeuristic` with the pattern collection (see `src/search/pdbs/pattern_collection_generator_hillclimbing.cc`). As an example for modifying the options, consider `handle_shrink_limit_options_defaults` in `src/search/merge_and_shrink/merge_and_shrink_algorithm.cc` where some options are set based on other options.

Reply:
Regarding 1: Static validation makes sense to me here; dynamic validation would not (e.g. things that depend on the task). I think these examples work as long as the only bools and lists we allow are *literals*. If we intend to support something like "let(pattern, something_that_generates_a_list_of_variables(), static_pdb(pattern))" in the future, i.e., make lists "full" types that can be returned by plug-ins and bound to variables, I don't think we can do this kind of validation any more because we cannot know without running the code whether the list is empty. (Same as: we cannot validate in C++ at compile-time, for example via a static_assert, whether a list passed through a variable is empty, but we could for list constants.) I think you brought up the idea of binding lists to a variable, so it's perhaps something we should discuss. For dynamic validation, I think it perhaps belongs in the builder, not in the plug-in?

For 2, this could also be at the builder level. I think the question here is whether certain kinds of rewrites should be macros (= happen at "compile time") or functions (happen at "runtime"). I think there are pros and cons for both, primarily having to do with diagnostics and logging output. Perhaps worth a discussion. I'm sympathetic to the idea of being able to implement things like ipdb() or astar() essentially as macros because, if we support this well, this could perhaps lead to the least boiler-plate-y implementation. But I think to be really useful, there would be a substantial one-time implementation cost. Right now I'd lean more towards a "put it in the factory" model, but it's something we should perhaps discuss.

Reply:
I don't think any of the validation depends on the task, so it is all static in that sense. It is called just before creating the objects, after all nested objects are already created. In programming terms: after we evaluated the parameters of a function call but before calling the function.

For the second case, I agree that putting the functionality in the builder or implementing it as a macro would both be valid options but I see this as beyond the scope for this issue as it is related to the builders that should yet not exist here.

Reply:
> I don't think any of the validation depends on the task.

Sure, I understand this is currently the case. My main point was that this would no longer be the case if we allowed binding lists to variables, which I think was something you mentioned earlier this week as an option. I'm fine with not going into the direction of dynamic lists, it's just a decision that we should make consciously.

Reply:
We discussed this in person and decided that splitting the functionality a bit differently handles both problems. Instead of preprocessing the options and then always instantiating the constructed object with the result, the derived plugin now is responsible for returning a shared pointer to the constructed object.